### ISSUE TITLE
Weak comparison detection

### ISSUE THREAD

**Issue by Soullivaneuh** [phpstan/phpstan-strict-rules#8]
Related: https://github.com/phpstan/phpstan/issues/791

A weak comparison error should be thrown for such case:

```php
<?php declare(strict_type=1);

function getCount(): ?int
{
	return 0 === rand(1, 5) % 5 ? 42 : null;
}

if (!getCount()) {
	echo 'test';
}
```

Because `0` and `null` will go to the same result.

But this case **must not** rise an error:

```php
 <?php declare(strict_type=1);

function getUpdatedAt(): ?\DateTime
{
	return 0 === rand(1, 5) % 5 ? new \DateTime('now') : null;
}

if (!getUpdatedAt()) {
	echo 'test';
}
```

Because it's currently fine, no confusable value can be return here.

**Comment by ondrejmirtes** [comment]
I've just written in the original issue: "there's already enforced that everything after ! must be a boolean."

But there's no harm in reporting both when all the rules are included so this is valid and will be accepted 😊

**Comment by Soullivaneuh** [comment]
Well, as I said in https://github.com/phpstan/phpstan/issues/791#issuecomment-359809268, I just don't see the value to force only boolean for `!` operator. But I may be missing something... Reference and/or sample would be great on the documentation IMHO.
